src/clean-core/to_string.cc

```diff
@@ -473,265 +473,159 @@ char const* unsigned_int_to_binary(IntType value, char* end)
 template <class IntType>
 void unsigned_to_string_impl(cc::string_stream_ref ss, IntType value, parsed_fmt_args const& args)
 {
+    auto const add_fill = [&](int length) {
+        auto const fill = args.sign_aware_zero_padding ? '0' : args.fill;
+        if (args.width > 0 && length < args.width)
+            for (auto i = 0; i < args.width - length; ++i)
+                ss << fill;
+    };
+
+    auto const to_binary = [&]() {
+        char buffer[sizeof(IntType) * 8 + 1];
+        char* end = buffer + sizeof(buffer);
+        char const* begin = unsigned_int_to_binary(value, end);
+        ss << cc::string_view(begin, end);
+    };
+
     static_assert(std::is_unsigned_v<IntType>);
     switch (args.type)
     {
     case 0:   // fallthrough
     case 'd': // default, decimal
     {
+        auto const buffer_size = [] {
+            switch (sizeof(IntType))
+            {
+            case 1:
+                return 3 + 1; // 255
+            case 2:
+                return 6 + 1; // 65535
+            case 4:
+                return 10 + 1; // 4 294 967 295
+            case 8:
+                return 20 + 1; // 18 446 744 073 709 551 615
+            }
+        }();
+
+        char buffer[buffer_size];
+        int length = 0;
         if constexpr (sizeof(IntType) == 1)
-        {
-            char buffer[3 + 1 + 1];
-            auto const length = std::snprintf(buffer, sizeof(buffer), "%hhu", value);
-            ss << cc::string_view(buffer, length);
-        }
+            length = std::snprintf(buffer, sizeof(buffer), "%hhu", value);
         if constexpr (sizeof(IntType) == 2)
-        {
-            char buffer[6 + 1 + 1];
-            auto const length = std::snprintf(buffer, sizeof(buffer), "%hu", value);
-            ss << cc::string_view(buffer, length);
-        }
+            length = std::snprintf(buffer, sizeof(buffer), "%hu", value);
         if constexpr (sizeof(IntType) == 4)
-        {
-            char buffer[10 + 1 + 1];
             // the cast prevents a warning when unsigned long is 32bit
-            auto const length = std::snprintf(buffer, sizeof(buffer), "%u", static_cast<unsigned int>(value));
-            ss << cc::string_view(buffer, length);
-        }
+            length = std::snprintf(buffer, sizeof(buffer), "%u", static_cast<unsigned int>(value));
         if constexpr (sizeof(IntType) == 8)
-        {
-            char buffer[19 + 1 + 1];
             // the cast prevents a warning when unsigned long is 64bit
-            auto const length = std::snprintf(buffer, sizeof(buffer), "%llu", static_cast<unsigned long long>(value));
-            ss << cc::string_view(buffer, length);
-        }
+            length = std::snprintf(buffer, sizeof(buffer), "%llu", static_cast<unsigned long long>(value));
```

Comment:
could this be simplified to a single printf with %llu, unconditionally casting to 64b? or at least the two smaller types.

Reply:
It *could* be at the at the cost of efficiency. Hypothetically though, because I did not measure the difference.

Reply:
On a side-note: I plan on getting rid of printf on the long run, which should simplify this anyways.

Reply:
For the smaller types it doesn't make a difference: https://godbolt.org/z/s13T89
iirc smaller types get promoted to int size in varargs anyway, just like float is always converted to double as well. Those %hhu conversions seem to just exist for symmetry with fscanf

Reply:
Ah, good to know. I new it does that for float -> double, but [cppreference](https://en.cppreference.com/w/cpp/io/c/fprintf) seems to suggest that there is a difference for different integer types.